src/cloudant/result.py

```diff
@@ -176,99 +217,193 @@ class Result(object):
+            :class:`~cloudant.result.ResultByKey`.
+
+        :returns: Rows data as a list in JSON format
         """
-        if isinstance(key, STRTYPE):
-            data = self._ref(key=key, **self.options)
-            return self._parse_data(data)
-
-        if isinstance(key, list):
-            data = self._ref(key=key, **self.options)
-            return self._parse_data(data)
-
-        if isinstance(key, slice):
-            # slice is startkey and endkey if str or array
-            str_or_none_start = type_or_none((STRTYPE, list), key.start)
-            str_or_none_stop = type_or_none((STRTYPE, list), key.stop)
-            if str_or_none_start and str_or_none_stop:
-                # startkey/endkey
-                if key.start is not None and key.stop is not None:
-                    data = self._ref(
-                        startkey=key.start,
-                        endkey=key.stop,
-                        **self.options
-                    )
-                if key.start is not None and key.stop is None:
-                    data = self._ref(startkey=key.start, **self.options)
-                if key.start is None and key.stop is not None:
-                    data = self._ref(endkey=key.stop, **self.options)
-                if key.start is None and key.stop is None:
-                    data = self._ref(**self.options)
-                return self._parse_data(data)
-            # slice is skip:skip+limit if ints
-            int_or_none_start = type_or_none(int, key.start)
-            int_or_none_stop = type_or_none(int, key.stop)
-            if int_or_none_start and int_or_none_stop:
-                if key.start is not None and key.stop is not None:
-                    limit = key.stop - key.start
-                    data = self._ref(
-                        skip=key.start,
-                        limit=limit,
-                        **self.options
-                    )
-                if key.start is not None and key.stop is None:
-                    data = self._ref(skip=key.start, **self.options)
-                if key.start is None and key.stop is not None:
-                    data = self._ref(limit=key.stop, **self.options)
-                # both None case handled already
-                return self._parse_data(data)
-        msg = (
-            'Failed to interpret the argument {0} '
-            'as a valid key value or as a valid slice.'
-        ).format(key)
-        raise CloudantArgumentError(msg)
+        data = None
+        if isinstance(arg, int):
+            data = self._handle_result_by_index(arg)
+        elif isinstance(arg, STRTYPE) or isinstance(arg, list):
+            data = self._handle_result_by_key(arg)
+        elif isinstance(arg, ResultByKey):
+            data = self._handle_result_by_key(arg())
+        elif isinstance(arg, slice):
+            # slice is entire result set - no additional processing required
```

Comment:
What this comment means is that the slice requested is the entire result set.  So we don't need to do any additional processing to the request as we do for key slicing or index slicing.